### PR TITLE
in region, treat current (and future) item-likes alike

### LINKED ISSUE #37864: Compile-Time Performance Regression
#37660 appears to have regressed performance by ~6% on bootstrap, due to a near tripling in time for item-bodies checking (23s to 62s). I'm not sure if that was expected or not, but someone should probably investigate. Let me know if I should open a new issue about that.

See [here](http://perf.rust-lang.org/compare.html?date_a=2016-10-19T05%3A13%3A09.000Z&date_b=2016-11-18T01%3A31%3A01.000Z&kind=benchmarks&crates=futures-rs-test-all%2Chelloworld%2Chtml5ever-2016-08-25%2Chyper.0.5.0%2Cinflate-0.1.0%2Cissue-32062-equality-relations-complexity%2Cissue-32278-big-array-of-strings%2Cjld-day15-parser%2Cpiston-image-0.10.3%2Cpiston-image-0.3.11%2Cregex-macros.0.1.30%2Cregex.0.1.30%2Crust-encoding-0.3.0%2Crust-encoding.0.2.32%2Csyntex-0.42.2%2Csyntex-0.42.2-incr-clean&phases=total&group_by=crate) for a comparison across all crates.

cc @nikomatsakis

### PR DESCRIPTION
**Pull request by nikomatsakis**
The `visit_fn` code mutates its surrounding context.  Between *items*,
this was saved/restored, but between impl items it was not. This meant
that we wound up with `CallSiteScope` entries with two parents (or
more!).  As far as I can tell, this is harmless in actual type-checking,
since the regions you interact with are always from at most one of those
branches. But it can slow things down.

Before, the effect was limited, since it only applied to impl items
within an impl. After #37660, impl items are visisted all together at
the end, and hence this could create a very messed up
hierarchy. Isolating impl item properly solves both issues.

I cannot come up with a way to unit-test this; for posterity, however,
you can observe the messed up hierarchies with a test as simple as the
following, which would create a callsite scope with two parents both
before and after

```
struct Foo {
}

impl Foo {
    fn bar(&self) -> usize {
        22
    }

    fn baz(&self) -> usize {
        22
    }
}

fn main() { }
```

Fixes #37864.

r? @michaelwoerister 

cc @pnkfelix -- can you think of a way to make a regr test?

### CODE REVIEW

**Comment by arielb1** [comment]
For example, the following lint could do:
```Rust
#![feature(plugin_registrar, rustc_private)]

extern crate syntax;
#[macro_use]
extern crate rustc;
extern crate rustc_plugin;

use rustc::lint::{LateContext, LintPass, LateLintPass, LintArray, LintContext};
use rustc::hir;
use rustc::hir::intravisit::FnKind;
use rustc::middle::region::CodeExtent;
use rustc::util::nodemap::FnvHashMap;

use syntax::ast::{self, NodeId};
use syntax::codemap::Span;

declare_lint!(REGION_HIERARCHY, Warn, "warn about bogus region hierarchy");

struct Pass {
    map: FnvHashMap<CodeExtent, NodeId>
}

impl LintPass for Pass {
    fn get_lints(&self) -> LintArray { lint_array!(REGION_HIERARCHY) }
}

impl LateLintPass for Pass {
    fn check_fn(&mut self, cx: &LateContext,
                fk: FnKind, _: &hir::FnDecl, expr: &hir::Block,
                span: Span, node: ast::NodeId)
    {
        if let FnKind::Closure(..) = fk { return }

        let mut extent = cx.tcx.region_maps.node_extent(expr.id);
        while let Some(parent) = cx.tcx.region_maps.opt_encl_scope(extent) {
            extent = parent;
        }
        if let Some(other) = self.map.insert(extent, node) {
            cx.span_lint(REGION_HIERARCHY, span, &format!(
                "different fns {:?}, {:?} with the same root extent {:?}",
                cx.tcx.map.local_def_id(other),
                cx.tcx.map.local_def_id(node),
                extent));
        }
    }
}

#[plugin_registrar]
pub fn plugin_registrar(reg: &mut ::rustc_plugin::Registry) {
    reg.register_late_lint_pass(Box::new(
        Pass { map: FnvHashMap() }
    ));
}
```

**Comment by michaelwoerister** [comment]
The fix looks good to me. Do you still want to experiment with adding a test case for this?

**Comment by nikomatsakis** [comment]
> The fix looks good to me. Do you still want to experiment with adding a test case for this?

I'm torn. The idea of a custom lint is clever. But it seems like a lot of effort for a data structure I would hope that we will overhaul in the not that distant future...

**Comment by nikomatsakis** [comment]
Well, @arielb1's lint basically works right off, so I might as well add it =)

**Comment by nnethercote** [comment]
The CI is failing due to some C++ compile error involving LLVM that I don't understand.

File: src/test/run-pass/issue-37290/main.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// aux-build:lint.rs
+
+#![feature(plugin)]
+#![plugin(lint)]
```
Comment (nikomatsakis) [inline src/test/run-pass/issue-37290/main.rs]: I suppose I do...

**Comment by michaelwoerister** [comment]
tidy error: 
```
C:\bot\slave\auto-win-gnu-32-opt-rustbuild\build\src\librustc\middle\region.rs:1069: line longer than 100 chars
```

**Comment by nikomatsakis** [comment]
Bah humbug.

**Comment by arielb1** [comment]
```
---- [run-pass] run-pass/issue-37290/main.rs stdout ----
	
error: auxiliary build of "/buildslave/rust-buildbot/slave/auto-linux-musl-64-opt/build/src/test/run-pass/issue-37290/auxiliary/lint.rs" failed to compile: 
status: exit code: 101
command: x86_64-unknown-linux-gnu/stage2/bin/rustc /buildslave/rust-buildbot/slave/auto-linux-musl-64-opt/build/src/test/run-pass/issue-37290/auxiliary/lint.rs -L x86_64-unknown-linux-gnu/test/run-pass/ --target=x86_64-unknown-linux-musl --error-format json --crate-type=lib -L x86_64-unknown-linux-gnu/test/run-pass/issue-37290/main.stage2-x86_64-unknown-linux-musl.run-pass.libaux -C prefer-dynamic --out-dir x86_64-unknown-linux-gnu/test/run-pass/issue-37290/main.stage2-x86_64-unknown-linux-musl.run-pass.libaux -C linker=/musl-x86_64/bin/musl-gcc -C ar=ar --cfg rtopt -C rpath -O -L x86_64-unknown-linux-musl/rt -C prefer-dynamic
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
{"message":"can't find crate for `syntax`","code":{"code":"E0463","explanation":"\nA plugin/crate was declared but cannot be found. Erroneous code example:\n\n```compile_fail,E0463\n#![feature(plugin)]\n#![plugin(cookie_monster)] // error: can't find crate for `cookie_monster`\nextern crate cake_is_a_lie; // error: can't find crate for `cake_is_a_lie`\n```\n\nYou need to link your code to the relevant crate in order to be able to use it\n(through Cargo or the `-L` option of rustc example). Plugins are crates as\nwell, and you link to them the same way.\n"},"level":"error","spans":[{"file_name":"/buildslave/rust-buildbot/slave/auto-linux-musl-64-opt/build/src/test/run-pass/issue-37290/auxiliary/lint.rs","byte_start":651,"byte_end":671,"line_start":18,"line_end":18,"column_start":1,"column_end":21,"is_primary":true,"text":[{"text":"extern crate syntax;","highlight_start":1,"highlight_end":21}],"label":"can't find crate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":null}
```

**Pull request merged**